storage/filestore/dir.go

```diff
@@ -238,6 +348,136 @@ func (dir *Dir) GarbageCollect(ctx context.Context) (err error) {
 	return nil
 }
 
+const nameBatchSize = 1024
+
+// GetAllNamespaces finds all known namespace IDs in use in local storage. They are not
+// guaranteed to contain any blobs.
+func (dir *Dir) GetAllNamespaces(ctx context.Context) (ids [][]byte, err error) {
+	defer mon.Task()(&ctx)(&err)
+	topBlobDir := dir.blobsdir()
+	openDir, err := os.Open(topBlobDir)
+	if err != nil {
+		return nil, err
+	}
+	defer func() { err = errs.Combine(err, openDir.Close()) }()
+	for {
+		dirNames, err := openDir.Readdirnames(nameBatchSize)
+		if err != nil && err != io.EOF {
+			return nil, err
+		}
+		if len(dirNames) == 0 {
+			return ids, nil
+		}
+		for _, name := range dirNames {
+			namespace, err := pathEncoding.DecodeString(name)
+			if err != nil {
+				// just an invalid directory entry, and not a namespace. probably
+				// don't need to pass on this error
+				continue
+			}
+			ids = append(ids, namespace)
+		}
+	}
+}
+
+// ForAllKeysInNamespace executes doForEach for each locally stored blob, stored with storage
+// format V1 or greater, in the given namespace. If doForEach returns a non-nil error,
+// ForAllKeysInNamespace will stop iterating and return the error immediately.
+func (dir *Dir) ForAllKeysInNamespace(ctx context.Context, namespace []byte, doForEach func(storage.StoredBlobAccess) error) (err error) {
+	namespaceDir := pathEncoding.EncodeToString(namespace)
+	nsDir := filepath.Join(dir.blobsdir(), namespaceDir)
+	openDir, err := os.Open(nsDir)
+	if err != nil {
+		if os.IsNotExist(err) {
```

Comment:
Are we sure we want to swallow the error here? If I call `ForAlLKeysInNamespace` but provide a namespace that doesn't exist, I think I'd expect an error. But I can see it both ways, maybe it's fine. :smile:

Reply:
Yes- we expect a namespace with no blobs to manifest as a nonexistent directory. Our machinery already deletes a namespace directory when it notices we've deleted the last blob in it. And there isn't any way to "create" a namespace with zero blobs in it.